File: CrackingTheCodingInterview/Question1.c
//
//  Question1.c
//  CrackingTheCodingInterview
//
//  Created by Chase Midler on 8/26/15.
//  Copyright (c) 2015 cmidler. All rights reserved.



//  Implement an algorithm to determine if a string has all unique characters. What
//  if you cannot use additional data structures?

#include "Question1.h"

//only considering ascii characters and assuming a == A
bool allUniqueChars(char* string)
{
    int len = (int)strlen(string);
    if(len <=1)
        return true;
    
    
    char check[32];
    
    memset(check,0,32);
    
    
    for(int i =0; i < len; i++)
    {
        char c = string[i];
        int index = 0;
        if(c>='0' && c<='9')
            index = 25+c-'0';
        else if (c>='A' && c<='Z')
            index = c-'A';
        else if (c>='a' && c<='z')
            index = c-'a';
        else
            return false;
        
        if(check[index]==1)
            return false;
        else
            check[index] = 1;
    }
    
    
    return true;
}

File: CTCI-Java/src/Chapter1/Question1.java
package Chapter1;

/**
 * Created by cptmidler on 7/6/17.
 * Is Unique: Implement an algorithm to determine if a string has all unique characters. What if you
 cannot use additional data structures?
 */
import java.util.HashMap;
import java.util.Arrays;
public class Question1 {
    //O(n) space and time
    boolean hasUCharacters(String s) {
        HashMap h = new HashMap();
        for(int i = 0; i<s.length(); i++)
        {
            char c = s.charAt(i);
            if(h.containsKey(c))
                return false;
            h.put(c,1);
        }
        return true;
    }

    //assuming lowercase english letters only (26 values in 8 byte int)
    boolean betterHasUCharacters(String s){
        int check = 0;
        for(int i =0; i<s.length(); i++)
        {
            int val = s.charAt(i) - 'a';
            if((check & (1 <<val)) >0)
                return false;
            check |=(1<<val);
        }

        return true;
    }

    public static void main(String[] args) {
        System.out.println(new Question1().hasUCharacters("test"));
        System.out.println(new Question1().hasUCharacters("true"));
        System.out.println(new Question1().betterHasUCharacters("test"));
        System.out.println(new Question1().betterHasUCharacters("true"));
    }
}
